### PR TITLE
Final pre-alpha stabilization of: iter, ops, slice, collections

### PR DESCRIPTION
**Pull request by aturon**
Now that various language features have been rolled out and RFCs implemented, this PR takes a final stabilization pass over `iter`, `ops`, `slice` and `collections`.

Commit summaries copied below:
### Final alpha stabilization of core::iter

This commit wraps up the adjustments to the iterator for recent language
changes.
- Moves `rposition` from `ExactSizeIterator` to `IteratorExt` using a
  `where` clause, thereby removing the `ExactSizeIterator:
  DoubleEndedIterator` constraint.
- Merges `MutableDoubleEndedIterator` into `IteratorExt`, renaming
  `reverse_` to `reverse_in_place`.
- Merges `IteratorOrdExt`, `IteratorCloneExt` and `CloneIteratorExt`
  into `IteratorExt` using `where` clauses.

Marks as `#[stable]`:
- the `iter` module itself
- `FromIterator`, `Extend`
- `Iterator`, `IteratorExt`
- `map`
- `filter`
- `filter_map`
- `skip_while`
- `take_while`
- `scan`
- `flat_map`
- `inspect`
- `collect`
- `fold`
- `all`
- `any`
- `find`
- `rposition`
- `max`, `min`
- Various adapter types related to the above methods

Because of the trait merging, this is a:

[breaking-change]
### Stabilize core::ops

This commit marks as stable those parts of `core::ops` that are in their
final planned form: `Drop`, all of the mathematical operators (`Add`,
`Sub`, etc), `Deref`/`DerefMut`. It leaves the `Index*`, `Slice*` and
`Fn*` traits unstable, as they are still undergoing active changes.
### Final alpha stabilization of std::slice

Marks as `#[stable]`:
- Various iterator structs for stable methods, e.g. `Chunks` and
  `Windows`.
- The `SliceExt` trait itself.
### Stabilize collection modules

The earlier collections stabilization did not cover the modules
themselves. This commit marks as stable those modules whose types have
been stabilized.
### Stabilization of impls and fallout from stabilization

Stabilizes `impl`s of now-stable traits.

### CODE REVIEW

**Comment by aturon** [comment]
(We should probably wait to land this until we've had a chance to discuss `Iterator` again.)

File: src/liballoc/arc.rs [hunk 1]
```diff
@@ -418,7 +418,7 @@ impl<T: Sync + Send> Clone for Weak<T> {
 }
 
 #[unsafe_destructor]
-#[experimental = "Weak pointers may not belong in this module."]
+#[stable]
```
Comment (Gankro) [inline src/liballoc/arc.rs]: Any notes on this call?

Comment (aturon) [inline src/liballoc/arc.rs]: Don't read anything into the `#[stable]` here other than: *if* we have `Weak`, we'll keep the `Drop` impl.

File: src/libstd/collections/hash/map.rs [hunk 2]
```diff
@@ -10,6 +10,8 @@
 //
 // ignore-lexer-test FIXME #15883
 
+#![stable]
```
Comment (Gankro) [inline src/libstd/collections/hash/map.rs]: This is a bit odd, since these modules aren't actually publically visible?

Comment (aturon) [inline src/libstd/collections/hash/map.rs]: They're re-exported.

File: src/libstd/collections/mod.rs [hunk 3]
```diff
@@ -309,7 +309,7 @@
 //! }
 //! ```
 
-#![experimental]
+#![stable]
```
Comment (Gankro) [inline src/libstd/collections/mod.rs]: Should we not be marking some of the facade modules made in this file (and/or libcollections/lib.rs) as #stable?

File: src/libstd/collections/hash/map.rs (hunk 2 above)
Comment (Gankro) [inline src/libstd/collections/hash/map.rs]: Well, we create a new module and pub use * in them, I'm not sure if that counts? e.g: https://github.com/rust-lang/rust/blob/master/src/libcollections/lib.rs#L77

File: src/libstd/collections/mod.rs (hunk 3 above)
Comment (aturon) [inline src/libstd/collections/mod.rs]: Ooops, you're right, the `pub use self::hash_map::HashMap` threw me off here. This PR does include stabilizations of the facades in `libcollections/lib.rs`, though.

**Comment by aturon** [comment]
@Gankro pushed a revision marking the correct `hash_map` modules. I think the markers within `libcollections` were already OK. Double-check my work?

File: src/libcore/iter.rs [hunk 4]
```diff
@@ -91,26 +95,80 @@ pub trait Iterator {
...
     fn size_hint(&self) -> (uint, Option<uint>) { (0, None) }
 }
 
 /// Conversion from an `Iterator`
-#[unstable = "may be replaced by a more general conversion trait"]
+#[stable]
 pub trait FromIterator<A> {
     /// Build a container with elements from an external iterator.
     fn from_iter<T: Iterator<Item=A>>(iterator: T) -> Self;
 }
```
Comment (alexcrichton) [inline src/libcore/iter.rs]: This seems slightly odd to still have a type parameter now that `Item` is an associated type, but I suppose `T: Iterator` doesn't tell you much and you need to write down the `Item` type somewhere.

Just confirming.

File: src/libcore/iter.rs [hunk 5]
```diff
@@ -660,6 +665,145 @@ pub trait IteratorExt: Iterator + Sized {
         None
     }
 
+    /// Return the index of the last element satisfying the specified predicate
+    ///
+    /// If no element matches, None is returned.
+    #[inline]
+    #[stable]
+    fn rposition<P>(&mut self, mut predicate: P) -> Option<uint> where
```
Comment (alexcrichton) [inline src/libcore/iter.rs]: I'm personally not a huge fan of the `where`-by-itself (I like the next line), but oh well!

**Comment by alexcrichton** [comment]
This looks great to me, this is all really falling into place!

**Comment by Gankro** [comment]
Also should we explicitly mark the other modules as `unstable` with a reason?

**Comment by alexcrichton** [comment]
This should also add `#![stable]` to the `slice` module itself.

**Comment by aturon** [comment]
@alexcrichton updated. After a bunch of discussion on the topic, I think we should move forward stabilizing `Iterator` as in this PR. r?

**Comment by alexcrichton** [comment]
Looks like rust-installer was reverted, but r=me otherwise

File: src/libcore/iter.rs (hunk 4 above)
Comment (aturon) [inline src/libcore/iter.rs]: This is desired to allow types to be created (or extended) via iterators from multiple items. For example, `String` can be extended by `char` or `&str` iterators.

**Comment by aturon** [comment]
I hate submodules :anger:

**Comment by alexcrichton** [comment]
Landed in #20610

**Pull request closed without merging**